### PR TITLE
feat: add notifier to capture trades with other players

### PR DESCRIPTION
**Pull request by iProdigy**
Requested on discord; screenshots completed trades

image

### CODE REVIEW

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java [hunk 1]
```diff
@@ -290,9 +292,13 @@ private MultipartBody createBody(NotificationBody<?> mBody, @Nullable RequestBod
      * @apiNote scalePercent should be in (0, 1]
      * @implNote the image format is either "png" (lossless) or "jpeg" (lossy), both of which can be used in MIME type
      */
-    private static CompletableFuture<Map.Entry<String, byte[]>> captureScreenshot(DrawManager drawManager, double scalePercent) {
+    private static CompletableFuture<Map.Entry<String, byte[]>> captureScreenshot(DrawManager drawManager, double scalePercent, @Nullable Image screenshotOverride) {
```
Comment (pajlada) [inline src/main/java/dinkplugin/message/DiscordMessageHandler.java:295]: Update comment with new param

File: src/main/java/dinkplugin/notifiers/TradeNotifier.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    public void onWidgetLoad(WidgetLoaded event) {
+        if (event.getGroupId() == TRADE_CONFIRMATION_GROUP) {
+            boolean chatHidden = Utils.hideWidget(config.screenshotHideChat(), client, ComponentID.CHATBOX_FRAME);
+            boolean whispersHidden = Utils.hideWidget(config.screenshotHideChat(), client, PRIVATE_CHAT_WIDGET);
+            drawManager.requestNextFrameListener(frame -> {
+                image.set(frame);
+                Utils.unhideWidget(chatHidden, client, clientThread, ComponentID.CHATBOX_FRAME);
+                Utils.unhideWidget(whispersHidden, client, clientThread, PRIVATE_CHAT_WIDGET);
+            });
```
Comment (pajlada) [inline src/main/java/dinkplugin/notifiers/TradeNotifier.java:129]: We should make sure to test this with trade delays disabled, where both parties accept the trade as fast as possible

File: src/main/java/dinkplugin/notifiers/TradeNotifier.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    public void reset() {
+        image.lazySet(null);
+    }
+
+    public void onTradeMessage(String message) {
+        if (!TRADE_ACCEPTED_MESSAGE.equals(message) || !isEnabled()) {
+            this.reset();
+            return;
+        }
```
Comment (pajlada) [inline src/main/java/dinkplugin/notifiers/TradeNotifier.java:76]: Is there a risk of receiving an unrelated game message between taking the screenshot & the trade being accepted, causing the image to be reset?

File: README.md [hunk 4]
```diff
@@ -759,6 +760,49 @@ See [javadocs](https://static.runelite.net/api/runelite-api/net/runelite/api/Gra
...
+    "counterparty": "%COUNTERPARTY%",
+    "receivedItems": [
+      {
+        "id": 314,
+        "quantity": 100,
+        "priceEach": 2,
+        "name": "Feather"
+      }
+    ],
+    "discardedItems": [
```
Comment (pajlada) [inline README.md:790]: Using the phrasing from the trade dialog itself, `receivedItems` is correct, and we would want to use `givenItems` instead of `discardedItems`

**Review by pajlada — CHANGES_REQUESTED** [review]
Some small nitpicks/questions, overall looks good

File: README.md (hunk 4 above)
Comment (iProdigy) [reply README.md:790]: trying to think of alternatives, `givenItems` has the downside that "You were given item xyz" suggests receive... but congruence with widget may be worth

`relinquishedItems` seems too formal

switching to `boughtItems`/`soldItems` might suggest a marketplace like GE rather than direct barter

File: src/main/java/dinkplugin/notifiers/TradeNotifier.java (hunk 3 above)
Comment (iProdigy) [reply src/main/java/dinkplugin/notifiers/TradeNotifier.java:76]: This shouldn't be an issue; we are only looking at `TRADE` message types so normal game messages won't matter

`TRADE` includes:
```
Player wishes to trade with you.
Sending trade offer...
Player is busy at the moment.
Other player declined trade.
Accepted trade.
```

(and when the trade interface is open, "wishes to trade with you" messages do not occur since the player is busy)

File: src/main/java/dinkplugin/notifiers/TradeNotifier.java (hunk 2 above)
Comment (iProdigy) [reply src/main/java/dinkplugin/notifiers/TradeNotifier.java:130]: That edge case would probably require playing at very low fps to realistically reproduce

Comment (Felanbird) [reply src/main/java/dinkplugin/notifiers/TradeNotifier.java:130]: works fine _without_ very low FPS from what I can test

**Review by Felanbird — APPROVED** [review]
also tested a boring 10gp minimum while giving 9gp, worked fine

```
{
   "type":"TRADE",
   "playerName":"felanbird1",
   "accountType":"NORMAL",
   "dinkAccountHash":"506af75ee4c2848d191f7c110039950d5205531d7140e5d0d1e4315f",
   "clanName":"P i c k l e",
   "seasonalWorld":false,
   "extra":{
      "counterparty":"Felanbird",
      "receivedItems":[
         
      ],
      "givenItems":[
         {
            "id":12625,
            "quantity":1,
            "priceEach":10440,
            "name":"Stamina potion(4)"
         },
         {
            "id":995,
            "quantity":10,
            "priceEach":1,
            "name":"Coins"
         },
         {
            "id":453,
            "quantity":1,
            "priceEach":167,
            "name":"Coal"
         },
         {
            "id":12629,
            "quantity":1,
            "priceEach":5352,
            "name":"Stamina potion(2)"
         },
         {
            "id":454,
            "quantity":1,
            "priceEach":167,
            "name":"Coal"
         },
         {
            "id":536,
            "quantity":1,
            "priceEach":2471,
            "name":"Dragon bones"
         },
         {
            "id":12378,
            "quantity":1,
            "priceEach":1400,
            "name":"Adamant cane"
         },
         {
            "id":7946,
            "quantity":1,
            "priceEach":314,
            "name":"Monkfish"
         },
         {
            "id":22124,
            "quantity":1,
            "priceEach":11176,
            "name":"Superior dragon bones"
         },
         {
            "id":12782,
            "quantity":1,
            "priceEach":2123,
            "name":"Senntisten teleport"
         }
      ],
      "receivedValue":0,
      "givenValue":33620
   },
```
```
{
   "type":"TRADE",
   "playerName":"Felanbird",
   "accountType":"NORMAL",
   "dinkAccountHash":"d0162dc0b639e9a2e0693d9279d0f938b04d21d07026c33967fc7fca",
   "seasonalWorld":false,
   "extra":{
      "counterparty":"felanbird1",
      "receivedItems":[
         {
            "id":12625,
            "quantity":1,
            "priceEach":10440,
            "name":"Stamina potion(4)"
         },
         {
            "id":995,
            "quantity":10,
            "priceEach":1,
            "name":"Coins"
         },
         {
            "id":453,
            "quantity":1,
            "priceEach":167,
            "name":"Coal"
         },
         {
            "id":12629,
            "quantity":1,
            "priceEach":5352,
            "name":"Stamina potion(2)"
         },
         {
            "id":454,
            "quantity":1,
            "priceEach":167,
            "name":"Coal"
         },
         {
            "id":536,
            "quantity":1,
            "priceEach":2471,
            "name":"Dragon bones"
         },
         {
            "id":12378,
            "quantity":1,
            "priceEach":1400,
            "name":"Adamant cane"
         },
         {
            "id":7946,
            "quantity":1,
            "priceEach":314,
            "name":"Monkfish"
         },
         {
            "id":22124,
            "quantity":1,
            "priceEach":11176,
            "name":"Superior dragon bones"
         },
         {
            "id":12782,
            "quantity":1,
            "priceEach":2123,
            "name":"Senntisten teleport"
         }
      ],
      "givenItems":[
         
      ],
      "receivedValue":33620,
      "givenValue":0
   },
```

**Review by pajlada — APPROVED** [review]
